hvsock_test.go

```diff
@@ -0,0 +1,627 @@
+			sa.fromRaw(&ra)
+			if sa != tt.giveConn.remote {
+				t.Errorf("%s remote addr give: %v; want: %v", tt.name, sa, tt.giveConn.remote)
+			}
+		}
+	})
+}
+
+func TestHvSockReadWrite(t *testing.T) {
+	u := newUtil(t)
+	l, addr := serverListen(u)
+	tests := []struct {
+		req, rsp string
+	}{
+		{"hello ", "world!"},
+		{"ping", "pong"},
+	}
+
+	// a sync.WaitGroup doesnt offer a channel to use in a select with a timeout
+	// could use an errgroup.Group, but for now dual channels work fine
+	svCh := make(chan struct{})
+	go func() {
+		defer close(svCh)
+		c, err := l.Accept()
+		u.Must(err, "listener accept")
+		defer c.Close()
+
+		b := make([]byte, 64)
+		for _, tt := range tests {
+			n, err := c.Read(b)
+			u.Must(err, "server rx")
+
+			r := string(b[:n])
+			u.Assert(r == tt.req, fmt.Sprintf("server rx error: got %q; wanted %q", r, tt.req))
+
+			_, err = c.Write([]byte(tt.rsp))
+			u.Must(err, "server tx error, could not send "+tt.rsp)
+		}
+		n, err := c.Read(b)
+		u.Assert(n == 0, "server did not get EOF")
+		u.Is(err, io.EOF, "server did not get EOF")
+	}()
+
+	clCh := make(chan struct{})
+	go func() {
+		defer close(clCh)
+		cl, err := Dial(context.Background(), addr)
+		u.Must(err, "client  dial")
+		defer cl.Close()
+
+		b := make([]byte, 64)
+		for _, tt := range tests {
+			_, err := cl.Write([]byte(tt.req))
+			u.Must(err, "client tx error, could not send "+tt.req)
+
+			n, err := cl.Read(b)
+			u.Must(err, "client rx")
+
+			r := string(b[:n])
+			u.Assert(r == tt.rsp, fmt.Sprintf("client rx error: got %q; wanted %q", b[:n], tt.rsp))
+		}
+
+		u.Must(cl.CloseWrite())
+	}()
+
+	u.Wait(svCh, 15*time.Second, "server")
+	u.Wait(clCh, 15*time.Second, "client")
+}
+
+func TestHvSockReadTooSmall(t *testing.T) {
+	u := newUtil(t)
+	s := "this is a really long string that hopefully takes up more than 16 bytes ..."
+	l, addr := serverListen(u)
+
+	svCh := make(chan struct{})
+	go func() {
+		defer close(svCh)
+		c, err := l.Accept()
+		u.Must(err, "listener accept")
+		defer c.Close()
+
+		b := make([]byte, 16)
+		ss := ""
+		for {
+			n, err := c.Read(b)
+			if errors.Is(err, io.EOF) {
+				break
+			}
+			u.Must(err, "server rx")
+			ss += string(b[:n])
+		}
+
+		u.Assert(ss == s, fmt.Sprintf("got %q, wanted: %q", ss, s))
+	}()
+
+	clCh := make(chan struct{})
+	go func() {
+		defer close(clCh)
+		cl, err := Dial(context.Background(), addr)
+		u.Must(err, "client  dial")
+		defer cl.Close()
+
+		_, err = cl.Write([]byte(s))
+		u.Must(err, "client tx error, could not send")
+	}()
+
+	u.Wait(svCh, 15*time.Second, "server")
+	u.Wait(clCh, 15*time.Second, "client")
+}
+
+func TestHvSockCloseReadWriteListener(t *testing.T) {
+	u := newUtil(t)
+	l, addr := serverListen(u)
+
+	ch := make(chan struct{})
+	go func() {
+		defer close(ch)
+		c, err := l.Accept()
+		u.Must(err, "listener accept")
+		defer c.Close()
+
+		hv := c.(*HvsockConn)
+		//
+		// test CloseWrite()
+		//
+		_, err = c.Write([]byte(testStr))
```

Comment:
Additionally verify we wrote the whole payload?

Reply:
Well, the contract for an io.Writer states we need to return an error if we wrote < len(byteSlice), so if we got an error check the n is what I mean.